apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs

```diff
@@ -0,0 +1,178 @@
+        }
+
+        /// <summary>
+        /// Reports an exception that occurred to the Stackdriver Error Reporting API.
+        /// </summary>
+        public ReportErrorEventResponse Report(HttpContext context, Exception exception)
+        {
+            var errorEvent = CreateReportRequest(context, exception);
+            // If the client task has faulted this will throw when accessing 'Result'
+            return _clientTask.Result.ReportErrorEvent(_projectName, errorEvent);
+        }
+
+        /// <summary>
+        /// Gets information about the HTTP request and response when the exception occurred 
+        /// and populates a <see cref="HttpRequestContext"/> object.
+        /// </summary>
+        private HttpRequestContext CreateHttpRequestContext(HttpContext context)
+        {
+            HttpRequest request = context?.Request;
+            HttpResponse response = context?.Response;
+
+            return new HttpRequestContext()
+            {
+                Method = request?.Method?.ToString() ?? "",
+                Url = UriHelper.GetDisplayUrl(request) ?? "",
+                UserAgent = request?.Headers["User-Agent"] ?? "",
+                ResponseStatusCode = response?.StatusCode ?? 0,
+            };
+        }
+
+        /// <summary>
+        /// Gets information about the source location where the exception occurred 
+        /// and populates a <see cref="SourceLocation"/> object.
+        /// </summary>
+        private SourceLocation CreateSourceLocation(Exception exception)
+        {
+            if (exception == null)
+            {
+                return new SourceLocation();
+            }
+
+            StackTrace stackTrace = new StackTrace(exception, true);
+            StackFrame[] frames = stackTrace.GetFrames();
+            if (frames.Length == 0)
+            {
+                return new SourceLocation();
+            }
+
+            StackFrame frame = frames[0];
+            return new SourceLocation()
+            {
+                FilePath = frame.GetFileName() ?? "",
+                LineNumber = frame.GetFileLineNumber(),
+                FunctionName = frame.GetMethod()?.Name ?? "",
+            };
+        }
+
+        /// <summary>
+        /// Gets information about the exception that occurred and populates
+        /// a <see cref="ReportedErrorEvent"/> object.
+        /// </summary>
+        private ReportedErrorEvent CreateReportRequest(HttpContext context, Exception exception)
+        {
+            ErrorContext errorContext = new ErrorContext()
+            {
+                HttpRequest = CreateHttpRequestContext(context),
+                ReportLocation = CreateSourceLocation(exception)
+            };
+
+            return new ReportedErrorEvent()
+            {
+                Message = exception?.ToString() ?? "",
+                Context = errorContext,
+                ServiceContext = _serviceContext,
+            };
```

Comment:
Up to you here, but you can create the `ErrorContext` when setting the `Context` property and then turn the whole thing into an expression-bodied method.